Add support for using declarations and async using (#672)
dotnet/csharpstandard: standard/attributes.md

>
> *end example*

### §enumerator-cancellation The EnumeratorCancellation attribute

> Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). The value of this parameter is a combination of two cancellation tokens: the token which was actually passed to the method, and the token passed to the `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)` implementation on the generated iterator. The combined token will be canceled if either of the two source tokens are canceled.

Reviewer: "will be" suggests we're requiring that the compiler observes this attribute - but later we have "The compiler can create a combined token" - let's decide whether or not a C# 8 compiler *must* implement this.

Reviewer 2: @jskeet – I think we need to cover combining tokens, see [this comment](https://github.com/dotnet/csharpstandard/pull/672#pullrequestreview-3108512762) for choices what to include.

“the token which was actually passed to the method” is not clear to me, which is the method”? Not the one with a parameter with this annotation on it, as that would be circular, so which one is it?

Reviewer: It *is* "the method with the parameter with this annotation on it" - which I understand is surprising. (It's quite odd to pass an argument, and end up with a different value for the parameter... but that's what's happening here.)

Reviewer 3: Reworded to try and make this more clear.

> Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). When the body of async iterator method accesses this parameter, it will receive a combination of two cancellation tokens: the token which was actually passed to the method, and the token passed to the `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)` implementation on the generated iterator. The combined token will be canceled if either of the two source tokens are canceled.

Reviewer 2: > When the body of async iterator method accesses this parameter, it will receive a combination of two cancellation tokens

An unusual was to say “The value of the parameter will be…”.

> It is an error if the `EnumeratorCancellation` attribute is applied to more than one parameter. The compiler may produce a warning if the `EnumeratorCancellation` attribute is applied to a parameter of a different type than `CancellationToken` or on a method that isn't an asynchronous iterator (§15.15).

Reviewer 2: And if it is not applied to any parameter? The token passed to `GetAsyncEnumerator` is discarded, should that be Standardised?

Reviewer 3: More or less. It's not "discarded", in that the enumerator may use it. The iterator code won't have access to it. Added a clarifying sentence.

Reviewer 4: > It's not "discarded", in that the enumerator may use it.

It's not clear to me how the enumerator could use it besides to do the only thing it currently does with it, which is to pass it to the async method body. Perhaps, to skip calling the async method body entirely if the token is canceled before the first call to MoveNextAsync happens. But once the async method body starts running, what can the enumerator do in response to the token if the token hasn't been passed to the async method body? Inject the equivalent of ThrowIfCancellationRequested around every yield? I think that would be too surprising for the author of the async method to understand. If so, then enumerators really don't have an opportunity to do anything with the token besides pass it to the async method.

> It is an error if the `EnumeratorCancellation` attribute is applied to more than one parameter. The compiler may produce a warning if the `EnumeratorCancellation` attribute is applied to a parameter of a different type than `CancellationToken` or on a method that isn't an asynchronous iterator (§15.15).

Reviewer 4: ```suggestion
It is an error if the `EnumeratorCancellation` attribute is applied to more than one parameter. The compiler may produce a warning if the `EnumeratorCancellation` attribute is applied to a parameter of a different type than `CancellationToken` or on a method that isn't an asynchronous iterator (§15.15) which returns an asynchronous enumerable interface ([§15.15.3](classes.md#15153-enumerable-interfaces)) rather than an asynchronous enumerator interface ([§15.15.2](classes.md#15152-enumerator-interfaces)).
```

> It is an error if the `EnumeratorCancellation` attribute is applied to more than one parameter. The compiler may produce a warning if the `EnumeratorCancellation` attribute is applied to a parameter of a different type than `CancellationToken` or on a method that isn't an asynchronous iterator (§15.15) which returns an asynchronous enumerable interface ([§15.15.3](classes.md#15153-enumerable-interfaces)) rather than an asynchronous enumerator interface ([§15.15.2](classes.md#15152-enumerator-interfaces)). If no attributes have this parameter, the iterator won't have access to the `CancellationToken` argument for `GetAsyncEnumerator`.

Reviewer: The second half of this sentence is a bit hard to read - the "or on a method that isn't an asynchronous iterator which returns an asynchronous enumerable interface rather than an asynchronous enumerator interface" part.

I think it would be easier to read if we just removed "rather than an asynchronous enumerator interface".

Alternatively, we could change it to a set of bullet points, for situations of:

- The attribute is applied to the wrong type of parameter
- The attribute is applied on a method which isn't an asynchronous iterator
- The attribute is applied on a method which returns an asynchronous enumerator instead of an asynchronous enumerable interface

(But I'd prefer just deleting the "rather than an asynchonrous enumerator interface" part.)

Reviewer 2: Agree with @jskeet some wordsmithing needed. Also is not “If no attributes have this parameter” the wrong way round?

Reviewer 3: Wordsmithed in upcoming commit.

I did retain all the potential warnings. The committee should discuss if we want those to be normative.

> *Example*: The method `GetStringsAsync()` is an asynchronous iterator. Before doing any work to retrieve the next value, it checks the cancellation token to determine if the iteration should be canceled. If cancellation is requested, no further action is taken.
>
> <!-- Example: {template:"code-in-class-lib", name:"AsyncEnumeratorCancellation"} -->
> ```csharp
> static async IAsyncEnumerable<object> GetStringsAsync([EnumeratorCancellation] CancellationToken token)
> {
>     for (int i = 0; i < 10; i++)
>     {
>         if (token.IsCancellationRequested)
>         {
>             yield break; // Exit if cancellation is requested
>         }
>         await Task.Delay(100);
>         yield return i.ToString();
>     }
> }
> ```
>

> > The `EnumerationCancellation` attribute instructs the compiler that the associated parameter is used for cancellation. Other cancellation tokens should be composed into a single aggregate token and substituted for the argument to this parameter. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `WithCancellation` extension method. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the aggregate token.

Reviewer 2: > Other cancellation tokens should be composed into a single aggregate token and substituted for the argument to this parameter.

Unclear what *other* is here, there is one possibly passed to `GetAsyncEnumerator` are there others?

Note sure this example should be here, see the overall review comment.

Reviewer 3: One edit: update this example. Will address the overall review comment separatedly.

Reviewer 3: For the overall comment, this example did stay to show an example of when tokens would be combined.

Reviewer 4: Can we say "A cancellation token passed to GetAsyncEnumerator" so that readers have a definitive pointer to what this is talking about?

Reviewer 4: ~~Possibly also, can we clarify that EnumeratorCancellation only has an effect when the method's return type is `IAsyncEnumerable<T>` (thus, GetAsyncEnumerator exists) and not when the method's return type is `IAsyncEnumerator<T>`?~~

Moving this comment above.

> > The `EnumerationCancellation` attribute instructs the compiler that the associated parameter is used for cancellation. Other cancellation tokens should be composed into a single aggregate token and substituted for the argument to this parameter. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `WithCancellation` extension method. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the aggregate token.

Reviewer 4: "Combined" and "aggregate" were competing synonyms. "Combined" goes along with the .NET API call that you make to do the combining.

```suggestion
> The `EnumerationCancellation` attribute instructs the compiler that the associated parameter is used for cancellation. Other cancellation tokens should be composed into a single combined token and substituted for the argument to this parameter. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `WithCancellation` extension method. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the combined token.
```

> > The `EnumerationCancellation` attribute instructs the compiler that the associated parameter is used for cancellation. The cancellation token passed to `GetAsyncEnumerator` is combined with `sourceOne.Token` into a single combined token and substituted for the argument to this parameter. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `ToListAsync` extension method. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the combined token.

Reviewer: ```suggestion
> The `EnumerationCancellation` attribute instructs the compiler that the associated parameter is used for cancellation. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `ToListAsync` extension method. The cancellation token passed to `GetAsyncEnumerator` is combined with `sourceOne.Token` into a single combined token and substituted for the argument to this parameter. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the combined token.
```

(Just reordering the sentences, otherwise we're referring to `sourceOne.Token` without introducing the example code that contains it.)

Reviewer 2: This is in an example so maybe this is a quibble, but “The compiler can” is surely wrong. If token combination is in the Standard, as I expect it is, then in normative text the implementation *must*/*shall* combine the tokens, in informative descriptive text maybe the implementation *will* combine?

Reviewer 2: This is in an example so maybe this is a quibble, but “The compiler can” is surely wrong. If token combination is in the Standard, as I expect it is, then in normative text the implementation *must*/*shall* combine the tokens, in informative descriptive text maybe the implementation *will* combine?

>
> *end example*

### §enumerator-cancellation The EnumeratorCancellation attribute

Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). The argument for this parameter shall be combined with the argument passed to `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)`. This combined token shall be polled by `IAsyncEnumerator<T>.MoveNextAsync()` (§15.15.5.2). The tokens shall be combined into a single token as if by `CancellationToken.CreateLinkedTokenSource` and its `Token` property. The combined token will be canceled if either of the two source tokens are canceled. The combined token is seen as the argument to the asynchronous iterator method (§15.15) in the body of that method.

It is an error if the `EnumeratorCancellation` attribute is applied to more than one parameter. The compiler may produce a warning if:

> - The `EnumeratorCancellation` attribute is applied to a parameter of a different type than `CancellationToken`,

Reviewer 2: Language, maybe the following?

```suggestion
- The `EnumeratorCancellation` attribute is applied to a parameter of a type other than `CancellationToken`,
```

- or if the `EnumeratorCancellation` attribute is applied to a parameter on a method that returns an asynchronous enumerable interface ([§15.15.3](classes.md#15153-enumerable-interfaces)) rather than an asynchronous enumerator interface ([§15.15.2](classes.md#15152-enumerator-interfaces)).

The iterator won't have access to the `CancellationToken` argument for `GetAsyncEnumerator` when no attributes have this parameter.

> *Example*: The method `GetStringsAsync()` is an asynchronous iterator. Before doing any work to retrieve the next value, it checks the cancellation token to determine if the iteration should be canceled. If cancellation is requested, no further action is taken.
>
> <!-- Example: {template:"code-in-class-lib", name:"AsyncEnumeratorCancellation"} -->
> ```csharp
> static async IAsyncEnumerable<object> GetStringsAsync([EnumeratorCancellation] CancellationToken token)
> {
>     for (int i = 0; i < 10; i++)
>     {
>         if (token.IsCancellationRequested)
>         {
>             yield break; // Exit if cancellation is requested
>         }
>         await Task.Delay(100);
>         yield return i.ToString();
>     }
> }
> ```
>

> > The `EnumerationCancellation` attribute indicates that the associated parameter is used for cancellation. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `ToListAsync` extension method. The cancellation token passed to `GetAsyncEnumerator` is combined with `sourceOne.Token` into a single combined token and substituted for the argument to this parameter. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the combined token.

Reviewer 2: This uses `ToListAsync`, which looks like it wasn’t around in C# 8’s time (it is in .NET 10 and has been added to earlier versions as an add-on?), is there a more suitable choice? Using it and stating it is an extension method just seems to add noise to the example. (My own test code uses `GetAsyncEnumerator` directly along with `MoveNextAsync` & `while` – but that may not be “more suitable”!)

Reviewer 4: It was around in C# 8's time as the community library System.Linq.Async (https://www.nuget.org/packages/System.Linq.Async/4.0.0).

The library was merged into .NET 10 recently (https://learn.microsoft.com/en-us/dotnet/core/compatibility/core-libraries/10.0/asyncenumerable).

The reason behind my suggestion of ToListAsync was to be more representative of realistic use cases than WithCancellation or GetAsyncEnumerator, but there may be good reasons to pick a less common/more "contrived" example.

Reviewer 2: @jnm2 – Reading this from a Standard-reader perspective we have a sample which uses something not in the Standard (what does it do?), and tells us it is an extension method (wonders if that is significant?). This is of course a common challenge with examples.

FWIW this is a fragment from my, definitely contrived :-), testing code:

```csharp
CancellationTokenSource ctsA = new CancellationTokenSource();
CancellationTokenSource ctsB = new CancellationTokenSource();

IAsyncEnumerable<int> iae = GetIntsAsync(limit, ctsA.Token);
var iterator = iae.GetAsyncEnumerator(ctsB.Token);

while (await iterator.MoveNextAsync())
{
    int current = iterator.Current;
```

Now the reader might wonder why it doesn’t use a `foreach` of course, but overall doesn’t use anything other than what it or the Standard defines.

Reviewer 3: I added a definition for `ToListAsync` to the sample. That should address Nigel's concerns.

> > The `EnumerationCancellation` attribute indicates that the associated parameter is used for cancellation. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `ToListAsync` extension method. The cancellation token passed to `GetAsyncEnumerator` is combined with `sourceOne.Token` into a single combined token and substituted for the argument to this parameter. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the combined token.

Reviewer 2: The spec is to combine the tokens (as in line 1065), not “The compiler can”.

```suggestion
> The `EnumerationCancellation` attribute indicates that the associated parameter is used for cancellation. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `ToListAsync` extension method. The cancellation token passed to `GetAsyncEnumerator` is combined with `sourceOne.Token` into a single combined token and substituted for the argument to this parameter. This combined token requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the combined token.
```

The tense seems wrong to me as well – the token is polled, using “token requests” sounds like the token itself is the actor. But I’ve no wordsmithing to offer on this one.

Reviewer 4: I share the sense that it's not the token doing the requesting.

Fwiw, it's not just polling in the standard sense. It's relatively frequently that I use CancellationToken.Register which registers a callback to handle cancellation rather than polling the token.

I'd offer "signals" instead of "requests."

Reviewer 2: @jnm2 – It is just polling in the Standard sense as `Register` is not in the Standard.

I think “signals” is an improvement here (unless the reader is used to *nix where it implies something rather stronger than “requests”! :-))

Reviewer 3: I used "indicates". That seemed to flow well with the existing text.

The iterator won't have access to the `CancellationToken` argument for `GetAsyncEnumerator` when no attributes have this parameter.

> *Example*: The method `GetStringsAsync()` is an asynchronous iterator. Before doing any work to retrieve the next value, it checks the cancellation token to determine if the iteration should be canceled. If cancellation is requested, no further action is taken.
>
> <!-- Example: {template:"code-in-class-lib", name:"AsyncEnumeratorCancellation"} -->
> ```csharp
> static async IAsyncEnumerable<object> GetStringsAsync([EnumeratorCancellation] CancellationToken token)
> {
> await using (IAsyncEnumerator<string> enumerator = 
>     GetStringsAsync(sourceOne.Token).GetAsyncEnumerator(sourceTwo.Token));
> while (await enumerator.MoveNextAsync())
> {
>     string number = enumerator.Current;
>     if (number == "8") sourceOne.Cancel();
>     if (number == "5") sourceTwo.Cancel();
>     Console.WriteLine(number);
> }

> >

Reviewer 3: ```suggestion
> ```
> 
```

The iterator won't have access to the `CancellationToken` argument for `GetAsyncEnumerator` when no attributes have this parameter.

> *Example*: The method `GetStringsAsync()` is an asynchronous iterator. Before doing any work to retrieve the next value, it checks the cancellation token to determine if the iteration should be canceled. If cancellation is requested, no further action is taken.
>
> <!-- Example: {template:"code-in-class-lib", name:"AsyncEnumeratorCancellation"} -->
> ```csharp
> static async IAsyncEnumerable<object> GetStringsAsync([EnumeratorCancellation] CancellationToken token)
> {
> await using (IAsyncEnumerator<string> enumerator = 
>     GetStringsAsync(sourceOne.Token).GetAsyncEnumerator(sourceTwo.Token));
> while (await enumerator.MoveNextAsync())
> {
>     string number = enumerator.Current;
>     if (number == "8") sourceOne.Cancel();
>     if (number == "5") sourceTwo.Cancel();
>     Console.WriteLine(number);
> }
> ```

> > 

Reviewer 3: ```suggestion
>     yield return number;
> }
> ```
>
```

> > }

Reviewer 3: ```suggestion
> await using (IAsyncEnumerator<string> enumerator = 
>     GetStringsAsync(sourceOne.Token).GetAsyncEnumerator(sourceTwo.Token));
> while (await enumerator.MoveNextAsync())
> {
>     string number = enumerator.Current;
>     if (number == "8") sourceOne.Cancel();
>     if (number == "5") sourceTwo.Cancel();
>     Console.WriteLine(number);
> }
```

> ```csharp
> static async IAsyncEnumerable<object> GetStringsAsync([EnumeratorCancellation] CancellationToken token)
> {
>     for (int i = 0; i < 10; i++)
>     {
>         if (token.IsCancellationRequested)
>         {
>             yield break; // Exit if cancellation is requested
>         }
>         await Task.Delay(100);
>         yield return i.ToString();
>     }
> }
> ```
>
> The `EnumerationCancellation` attribute indicates that the associated parameter is used for cancellation. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `ToListAsync` extension method. The cancellation token passed to `GetAsyncEnumerator` is combined with `sourceOne.Token` into a single combined token and substituted for the argument to this parameter. This combined token that indicates that cancellation was requested when either token indicates cancellation. The async iterator then cancels the operation based on the state of the combined token.
>
> ```csharp
> CancellationTokenSource sourceOne = new CancellationTokenSource();
> CancellationTokenSource sourceTwo = new CancellationTokenSource();

> > await foreach (string number in GetStringsAsync(sourceOne.Token).ToListAsync(sourceTwo.Token))

Reviewer 2: @BillWagner & @jnm2 – I understand Joseph's wish to show something more realistic here, but I think it is hiding what the example needs to show - the connection between the asynchronous iterator and `GetAsyncEnumerator` where the language requires the combining of `CancellationToken`’s. To do this the example needs to show `GetAsyncEnumerator` so the description can refer to it, so back to something like:

```suggestion
> await foreach (string number in GetStringsAsync(sourceOne.Token).GetAsyncEnumerator(sourceTwo.Token))
```

which directly shows the two key items - the enumerable and the enumerator. Using `ToListAsync` just detracts from the purpose and adds complexity (and pitfalls as we've seen!)

Reviewer 3: @jnm2 @Nigel-Ecma Updated to use a modified version of Nigel's sample above (`await foreach` requires `IAsyncEnumerable`, not `IAsyncEnumerator`.

Reviewer 2: @BillWagner – Apologies, by bad, it was late and I foolishly translated from a while/move sample in my IDE to a foreach in the comment box…

### §enumerator-cancellation The EnumeratorCancellation attribute

> Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). The argument for this parameter shall be combined with the argument passed to `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)`. This combined token shall be polled by `IAsyncEnumerator<T>.MoveNextAsync()` (§15.15.5.2). The tokens shall be combined into a single token as if by `CancellationToken.CreateLinkedTokenSource` and its `Token` property. The combined token will be canceled if either of the two source tokens are canceled. The combined token is seen as the argument to the asynchronous iterator method (§15.15) in the body of that method.

Reviewer 4: From what I saw when I decompiled the compiler-generated code, there is no requirement that GetAsyncEnumerator polls the cancellation token. If the async method body ignores the cancellation token, then it is never observed.
```suggestion
Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). The argument for this parameter shall be combined with the argument passed to `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)`. The tokens shall be combined into a single token as if by `CancellationToken.CreateLinkedTokenSource` and its `Token` property. The combined token will be canceled if either of the two source tokens are canceled. The combined token is seen as the argument to the asynchronous iterator method (§15.15) in the body of that method.
```

Reviewer 4: In other words, [EnumeratorCancellation] is simply about making a merged token available to the method body, and is not about having the generated implementation of IAsyncEnumerable respond to any token.

> }
> ```
>
> The `EnumerationCancellation` attribute instructs the compiler that the associated parameter is used for cancellation. The cancellation token passed to `GetAsyncEnumerator` is combined with `sourceOne.Token` into a single combined token and substituted for the argument to this parameter. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `WithCancellation` extension method. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the combined token.
>
> ```csharp
> CancellationTokenSource sourceOne = new CancellationTokenSource();
> CancellationTokenSource sourceTwo = new CancellationTokenSource();
> await foreach (string number in GetStringsAsync(sourceOne.Token).WithCancellation(sourceTwo.Token))
> {
>     if (number == "8") sourceOne.Cancel();

> >     if (number == "5") sourceTwo.Cancel(); q

Reviewer 4: ```suggestion
>     if (number == "5") sourceTwo.Cancel();
```

>
> *end example*

### §enumerator-cancellation The EnumeratorCancellation attribute

> Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15).

Reviewer: At the moment it's not obvious to me why an attribute is needed. The description of the example says "Before doing any work to retrieve the next value, it checks the cancellation token to determine if the iteration should be cancelled" - that's part of the code in the method itself, but are we saying the compiler injects a check as well?

It may well be that reading 15.15 again will make this obvious, but I think we should expand the example anyway.

Reviewer 3: Added a couple clarifying sentences. Probably still more work needed.

(For example, what if the attribute is applied to a type that isn't a cancellation token?

Reviewer 4: The attribute causes the body of the method to receive a cancellation token that is not necessarily the one passed to the method by the method caller. Instead, the body of the method receives two cancellation tokens [stitched together](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.createlinkedtokensource): one is the token actually passed by the caller of the async iterator method, and the second is the token passed to [GetAsyncEnumerator(CancellationToken)](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.iasyncenumerable-1.getasyncenumerator).

```cs
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;

class Program
{
    static CancellationTokenSource cancellationTokenSource1 = new CancellationTokenSource();
    static CancellationTokenSource cancellationTokenSource2 = new CancellationTokenSource();

    public static async Task Main()
    {
        // Notice what happens if you pass CancellationToken.None for either one of these!
        await using var enumerator = IterateAsync(cancellationTokenSource1.Token).GetAsyncEnumerator(cancellationTokenSource2.Token);

        while (await enumerator.MoveNextAsync()) { }
    }

    public static async IAsyncEnumerable<int> IterateAsync([EnumeratorCancellation] CancellationToken cancellationToken)
    {
        Console.WriteLine(cancellationToken.Equals(cancellationTokenSource1.Token)); // False
        Console.WriteLine(cancellationToken.Equals(cancellationTokenSource2.Token)); // False
        yield break;
    }
}
```

The `WithCancellation` extension method returns a wrapper that passes a cancellation token to GetAsyncEnumerator. This example shows the more likely consumption pattern that results in a linked cancellation token:

```cs
    public static async Task Main()
    {
        // Notice what happens if you pass CancellationToken.None for either one of these!
        await foreach (var _ in IterateAsync(cancellationTokenSource1.Token).WithCancellation(cancellationTokenSource2.Token))
        {
        }
    }
```

See [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRhQG4BYAKGz3wDoAlAVwDsAXASwFsZqBhCdgB1YAbGFADKogG6swMAM6kKuAtQAqAC1gBDACatGAc0WUCAVkXkMAZjwAmHNxwBvcjjd4bBJHgAcOAJIAgnIAnoxgAKKM9JxQWsAiADz6zAB8OACywWFgABSu7oUA2lExolrM0Nxa4TBCQhWsEIwAug41svWNzaoQANYwjDhgHXUNbD39g/gANAWFbtW1XROMvQNDI8vjTWtTjLYAlPPOJ4UYAJx4SNQAInVaIbn4WDPDoyu769OHigvul2udweTxeby2nR2kw2Rz+/yo+BwwG0fThbgAvuR0UA)'s decompilation of the `GetAsyncEnumerator` method for the logic.

> (For example, what if the attribute is applied to a type that isn't a cancellation token?)

@BillWagner Only a warning:

> warning CS8424: The EnumeratorCancellationAttribute applied to parameter 'cancellationToken' will have no effect. The attribute is only effective on a parameter of type CancellationToken in an async-iterator method returning IAsyncEnumerable

Reviewer 4: We also have:

> error CS8426: The attribute [EnumeratorCancellation] cannot be used on multiple parameters

Reviewer 3: This should be addressed in an upcoming commit.

Reviewer 3: I did add an additional example, and some explanations. I'm concerned that the additions (especially the examples) are more "documenting an implementation" than should be in the standard. 

I'd like to mark this discussion for the meeting later this month.

> Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). The compiler generated iterator passes the argument for this parameter to the `IAsyncEnumerable<T>.GetAsyncEnumerator()` method. The iterator respects the token's state in the `MoveNextAsync` method and cancels the operation if possible.

Reviewer 3: The attribute is used for the `.WithCancellation` extension method. 

The compiler must wrap the original async iterator in a generated class that does enables the cancellation in the wrapper, not the wrapped state machine.

Reviewer 4: It's the other way around. The compiler-generated iterator actually _receives_ one of the inputs to is parameter by someone calling its implementation of `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)`.

Also, the iterator does not respect the token's state in MoveNextAsync. The token is only respected if the user explicitly uses the token in the body of their iterator method. (Verified both by example and by looking at generated code.)

```suggestion
Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). When the body of async iterator method accesses this parameter, it will receive a combination of two cancellation tokens: the token which was actually passed to the method, and the token passed to the `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)` implementation on the generated iterator. The combined token will be canceled if either of the two source tokens are canceled.
```

Reviewer 4: Yes, without calling either .WithCancellation (and explaining what it is?) or calling .GetAsyncEnumerator, the example doesn't demonstrate `[EnumeratorCancellation]`.

> Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). The compiler generated iterator passes the argument for this parameter to the `IAsyncEnumerable<T>.GetAsyncEnumerator()` method. The iterator respects the token's state in the `MoveNextAsync` method and cancels the operation if possible.

Reviewer 4: It's the other way around. The compiler-generated iterator actually _receives_ one of the inputs to its CancellationToken parameter by someone calling its implementation of `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)`.

Also, the iterator does not respect the token's state in MoveNextAsync. The token is only respected if the user explicitly respects the token by using it in the body of their iterator method. (Verified both by example and by looking at generated code.)

```suggestion
Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). When the body of async iterator method accesses this parameter, it will receive a combination of two cancellation tokens: the token which was actually passed to the method, and the token passed to the `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)` implementation on the generated iterator. The combined token will be canceled if either of the two source tokens are canceled.
```

Reviewer 4: And I bet @Nigel-Ecma is going to really love me throwing out the terms "token" and "canceled" without defining them 😁
But I do feel it's important to explain the behavior of the combined token, since that is the reason a user would want to depend on the feature.

Reviewer 3: I'll take the suggestion because it's a good starting point. I'll continue editing and clarify terms, or use terms already defined.

Reviewer 3: @jnm2 Can you post the example, or a link to a lab.razor.fyi example?

Reviewer 4: Sure thing! A runnable demo is at <https://github.com/dotnet/csharpstandard/pull/672#discussion_r2180979017> with a SharpLab link. I simplified it and moved it to lab.razor.fyi: [link](https://lab.razor.fyi/#dY_BSsRADIaxeNA5iU-Q4woy6LkiSBVREMTtSRFMs6GGTmeWyVQpss_lxZNP4sG7byDS3UV00f8S8vH_4Y_5yIy5jKGO2FrS7besU_E1jHtN3NoiOMeUJHi1p-w5CuXml-Oq80latkVop-I4jjk-CLGu2Mr7yDgRX__HbYnaaG7MtKucEJBDVSjgyQAALKEmTEKA2nuCs6NhnPiu5YiV4wPx6RAu5nQ0Tw26WRpSiAV6YudweOcWfm5laNgDrZKd-ZVFg0H4iJJgKGqP2WE_2t_b_SOVfwd6YTeBKjI2Czgzs_OM9Hpz4-X59f2Tt9bv1r4A)

You can see how the state machine implements GetAsyncEnumerator. GetAsyncEnumerator receives a second token, and then combines the two tokens if both are not `CancellationToken.None`:

<details>

```cs
[DebuggerHidden]
IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator(CancellationToken cancellationToken = default(CancellationToken))
{
    <MAsync>d__0 <MAsync>d__;
    if (<>1__state == -2 && <>l__initialThreadId == Environment.CurrentManagedThreadId)
    {
        <>1__state = -3;
        <>t__builder = AsyncIteratorMethodBuilder.Create();
        <>w__disposeMode = false;
        <MAsync>d__ = this;
    }
    else
    {
        <MAsync>d__ = new <MAsync>d__0(-3);
    }
    if (<>3__cancellationToken.Equals(default(CancellationToken)))
    {
        <MAsync>d__.cancellationToken = cancellationToken;
    }
    else if (cancellationToken.Equals(<>3__cancellationToken) || cancellationToken.Equals(default(CancellationToken)))
    {
        <MAsync>d__.cancellationToken = <>3__cancellationToken;
    }
    else
    {
        <>x__combinedTokens = CancellationTokenSource.CreateLinkedTokenSource(<>3__cancellationToken, cancellationToken);
        <MAsync>d__.cancellationToken = <>x__combinedTokens.Token;
    }
    return <MAsync>d__;
}
```

</details>

> > *Example*: The method `GetStringsAsync()` is an asynchronous iterator. Before doing any work to retrieve the next value, it checks the cancellation token to determine if the iteration should be cancelled. If Cancellation is requested, no further action is taken.

Reviewer: ```suggestion
> *Example*: The method `GetStringsAsync()` is an asynchronous iterator. Before doing any work to retrieve the next value, it checks the cancellation token to determine if the iteration should be cancelled. If cancellation is requested, no further action is taken.
```

>
> <!-- Example: {template:"code-in-class-lib", name:"AsyncEnumeratorCancellation"} -->
> ```csharp

> > static async IAsyncEnumerable<object> GetStringsAsync([EnumeratorCancellation]CancellationToken token)

Reviewer: ```suggestion
> static async IAsyncEnumerable<object> GetStringsAsync([EnumeratorCancellation] CancellationToken token)
```

(I think there's usually a space before the type conventionally. I've only checked with the NotNullWhen example above though...)

>
> *end example*

> ### §enumerator-cancellation The EnumeratorCancellation attribute

Author: Once we've merged this, I'll do the following:

1. Add a pointer to this attribute to the bullet list in 22.5.1, "General," where we provide a summary of Reserved attributes.
2. Convert the hard-coded links to relative ones.

Reviewer 3: @RexJaeschke I think this has all been addressed? Or do these still need to be updated after merging this PR?

Author: 1. Still need to add a pointer. Can ignore 2.

Reviewer 3: Thanks Rex. Add in the final commit.

> Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). When the body of async iterator method accesses this parameter, it will receive a combination of two cancellation tokens: the token which was actually passed to the method, and the token passed to the `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)` implementation on the generated iterator. The combined token will be cancelled if either of the two source tokens are canceled.

Reviewer: ```suggestion
Specifies the parameter representing the `CancellationToken` for an asynchronous iterator (§15.15). When the body of async iterator method accesses this parameter, it will receive a combination of two cancellation tokens: the token which was actually passed to the method, and the token passed to the `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)` implementation on the generated iterator. The combined token will be canceled if either of the two source tokens are canceled.
```

(Or go in the other direction, but let's be consistent.)

It is an error if the `EnumeratorCancellation` attribute is applied to more than one parameter. The compiler may produce a warning if the `EnumeratorCancellation` attribute is applied to a parameter of a different type than `CancellationToken` or on a method that isn't an asynchronous iterator (§15.15).

> > *Example*: The method `GetStringsAsync()` is an asynchronous iterator. Before doing any work to retrieve the next value, it checks the cancellation token to determine if the iteration should be cancelled. If cancellation is requested, no further action is taken.

Reviewer: ```suggestion
> *Example*: The method `GetStringsAsync()` is an asynchronous iterator. Before doing any work to retrieve the next value, it checks the cancellation token to determine if the iteration should be canceled. If cancellation is requested, no further action is taken.
```

> ```csharp
> static async IAsyncEnumerable<object> GetStringsAsync([EnumeratorCancellation] CancellationToken token)
> {
>     for (int i = 0; i < 10; i++)
>     {
>         if (token.IsCancellationRequested)
>         {
>             yield break; // Exit if cancellation is requested
>         }
>         await Task.Delay(100);
>         yield return i.ToString();
>     }
> }
> ```
>
> The `EnumerationCancellation` attribute instructs the compiler that the associated parameter is used for cancellation. Other cancellation tokens should be composed into a single aggregate token and substituted for the argument to this parameter. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `WithCancellation` extension method. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the aggregate token.
>
> ```csharp
> CancellationTokenSource sourceOne = new CancellationTokenSource();
> CancellationTokenSource sourceTwo = new CancellationTokenSource();

> > await foreach (string number in GetStringsAsync(sourceOne.Token).WithCancellation(sourceTwo.Token))

Reviewer: This may become obvious later but in case I forget: are we standardizing `WithCancellation`?

Reviewer 3: Decision from July 30 meeting: We'll use `WithCancellation` in examples, but not in normative text.

> }
> ```
>

> > The `EnumerationCancellation` attribute instructs the compiler that the associated parameter is used for cancellation. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `ToListAsync` extension method. The cancellation token passed to `GetAsyncEnumerator` is combined with `sourceOne.Token` into a single combined token and substituted for the argument to this parameter. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the combined token.

Reviewer 2: I first thought “instructs the compiler” could be replaced with “informs the compiler”, but that still talks about the compiler when this is surely just part of the language, so I ended up with:

```suggestion
> The `EnumerationCancellation` attribute indicates that the associated parameter is used for cancellation. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `ToListAsync` extension method. The cancellation token passed to `GetAsyncEnumerator` is combined with `sourceOne.Token` into a single combined token and substituted for the argument to this parameter. The compiler can create a combined token that requests cancellation when either token has requested cancellation. The async iterator then cancels the operation based on the state of the combined token.
```

> }
> ```
>
> The `EnumerationCancellation` attribute indicates that the associated parameter is used for cancellation. In the following example, one token is passed to the `GetStringsAsync` method, and another is passed to the `ToListAsync` extension method. The cancellation token passed to `GetAsyncEnumerator` is combined with `sourceOne.Token` into a single combined token and substituted for the argument to this parameter. This combined token that indicates that cancellation was requested when either token indicates cancellation. The async iterator then cancels the operation based on the state of the combined token.
>

> > ```csharp

Reviewer 2: @BillWagner – This sample is not compiling for me. Adding `async` to your version of `ToListAsync()` gets that method to compile, but overall it still fails at the `await foreach`

> CS8411 Asynchronous foreach statement cannot operate on variables of type `ValueTask<List<object>>` because `ValueTask<List<object>>` does not contain a suitable public instance definition for `GetAsyncEnumerator`

VS2022 (ARM), LangVersion=8.0, .NET 8

> ```csharp
> CancellationTokenSource sourceOne = new CancellationTokenSource();
> CancellationTokenSource sourceTwo = new CancellationTokenSource();
> IAsyncEnumerator<string> enumerator = GetStringsAsync(sourceOne.Token).GetAsyncEnumerator(sourceTwo.Token);
> while (await enumerator.MoveNextAsync())
> {
>     string number = enumerator.Current;
>     if (number == "8") sourceOne.Cancel();
>     if (number == "5") sourceTwo.Cancel();
>         Console.WriteLine(number);

> > }

Reviewer 4: Last nit from me. The foreach involves disposal, so when lowering `await foreach` down to a while loop, we should either use `await using` or a try...finally with explicit disposal.
```suggestion
> await using (IAsyncEnumerator<string> enumerator = GetStringsAsync(sourceOne.Token).GetAsyncEnumerator(sourceTwo.Token))
> {
>     while (await enumerator.MoveNextAsync())
>     {
>         string number = enumerator.Current;
>         if (number == "8") sourceOne.Cancel();
>         if (number == "5") sourceTwo.Cancel();
>         Console.WriteLine(number);
>     }
> }
```

>     Console.WriteLine(number);
> }
>
> public static ValueTask<List<TSource>> ToListAsync<TSource>(
>     this IAsyncEnumerable<TSource> source,
>     CancellationToken cancellationToken = default)
> {
>     List<TSource> list = new List<TSource>();

> >     await foreach (TSource element in source)

Reviewer 4: Importantly, ToListAsync here must use WithCancellation in order to cancel waiting for the next item. It could do this instead of directly polling the cancellation token. As written now, it will not signal the cancellation token in the async iterator method body.
